### ISSUE TITLE
Support for OpenShift OAuth Server for authentication and authorization

### ISSUE THREAD

**Issue by jtmulvey** [OpenLiberty/open-liberty#9457]
Open Liberty should support use of OpenShift OAuth 2.0 Server for authentication and authorization for Java EE, Jakarta, and Microprofile applications.  This feature would offer login re-direction in an OpenShift environment for web and REST API based logins, followed by the creation of a JWT to hold Group information for a User in the JWT claims.  
We will also add support in Liberty to accept Open Shift's OAuth tokens.

 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [x] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [x] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ n/a] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [n/a ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [x] Implementation complete. (Epic owner / Feature owner) 
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [x] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [x] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [x] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [x] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by jwalcorn** [comment]
Is this primarily "we should formally test this and document how to do it", or do we have reason to believe it wouldn't work today, without code changes in Liberty?  Consider that a preview question for the design review next week... ;>)

**Comment by jtmulvey** [comment]
@jwalcorn It doesn't work today. That's why we opened the issue. The OpenShift OAuth 2.0 server creates opaque tokens that are not supported today by our OAuth 2.0 implementation.

**Comment by chunlongliang-ibm** [comment]
@NottyCode  WAD has been uploaded to box for your approval.

**Comment by NottyCode** [comment]
@chunlongliang-ibm can you please update your comment to add a link to the design please?

**Comment by chunlongliang-ibm** [comment]
@NottyCode , link to the WAD, https://ibm.box.com/s/hfk7ntr81xh4zwizukqhi63t6vzao2ex

**Comment by chirp1** [comment]
@brutif Bruce indicated that no doc is needed. I'm approving.

**Comment by chunlongliang-ibm** [comment]
[WAD-EPIC9457-LibertyLoginWithOpenshiftOAuth.pdf](https://github.com/OpenLiberty/open-liberty/files/3856621/WAD-EPIC9457-LibertyLoginWithOpenshiftOAuth.pdf)

**Comment by chunlongliang-ibm** [comment]
@NottyCode The WAD has been converted to PDF, and attached to EPIC

**Comment by chunlongliang-ibm** [comment]
Will find how to make box link public

**Comment by marikaj123** [comment]
Epics going into the next release will need to contain the label target:ga

**Comment by marikaj123** [comment]
Does the Epic contain nls english files for translation?

**Comment by gscottj** [comment]
This feature does not have a UI that is exposed to end users or generate output that requires  accessibility verification testing. Approval granted.

**Comment by rumanaHaque** [comment]
SVT for this Feature is done using TER - 
SVT: TER: Test support for OpenShift OAuth Server for authentication and authorization (https://github.ibm.com/was-liberty/WS-CD-Open/issues/19430)
The TER is currently in progress. @meiaus (Mei-hsiang) is executing this TER.
The functionality of the TER has already been verified, and we had a review session where she went through the steps of the TER and the demo. 
Mei-hsiang is waiting for final build - to rerun the scenario to ensure everything works as expected. Once she completes the TER, I will approve the SVT record for this Feature.

**Comment by brutif** [comment]
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

    WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

    **The WAD notes that many errors are already handled by the parent feature (socialLogin-1.0).  New messages have been added to cover new errors that were not previously possible.**

    Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
    a) What problem paths were tested and demonstrated?  
    **Pretty much everything, misconfig of Openshift (both Oauth Server and proxy sidecar), misconfig of Liberty.**
    b) Who did you demo to?  
    **Test team** 
    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? 
    **TBD**

    SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?  Mei Hsiang Chang, Rumana Haque
    b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3? 
    **TBD**

    Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the following files and that the respective L2 / L3 teams know they are supporting it
    a) https://w3-03.ibm.com/systems/techlink/crf/viewRecord.do?category=CRF&uid=724&sessionTime=1514919054315&rsid=0&seqNo=2&displayId=724&searchStr=websphere+application+server&totalHits=30) - contact MICHAEL SCHNEIDER/Durham/IBM
    b) http://wasl3toolsvm.torolab.ibm.com:8080/QueueContactSummary - contact Eric M Covener/Durham/IBM or Raymond Koo/Toronto/IBM
    **WL3SEC, WL3WSS**

**Comment by rumanaHaque** [comment]
SVT testing is complete. Marked SVT approval.

**Comment by clarkek123** [comment]
For demo approval --- We were scheduled for a demo on 12/03/19, but had technical difficulties and the demo is now scheduled for 12/17/19.

**Comment by brutif** [comment]
Marking globalization approved for Sam Wong as he is having authorization problems.

**Comment by brutif** [comment]
New metatype attributes are now exposed in master with pull 10005.  GA blog is ready on issue 10047.  Closing this epic.
